cmd/zed/agents/zfs_retire.c

```diff
@@ -351,9 +351,8 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
 			zpool_vdev_offline(zhp, devname, B_TRUE);
 		} else if (!fmd_prop_get_int32(hdl, "spare_on_remove") ||
 		    replace_with_spare(hdl, zhp, vdev) == B_FALSE) {
-			/* Could not handle with spare: offline the device */
-			fmd_hdl_debug(hdl, "zpool_vdev_offline '%s'", devname);
-			zpool_vdev_offline(zhp, devname, B_TRUE);
+			/* Could not handle with spare: do not offline */
+			fmd_hdl_debug(hdl, "no spare for '%s'", devname);
```

Comment:
Perhaps this behavior should be conditional on the `autoreplace` property.  When `on` the device should be left in the `VDEV_STATE_REMOVED` state to support autoreplace.  When `off` the device can be set to `VDEV_STATE_OFFLINE`.  Note the `auto_offline_001_pos.ksh` test checks this behavior and will need to be updated accordingly.

Reply:
Well, I think UNAVAIL is the correct device state irrespective of the `autoreplace` property. After all, the device *really* is UNAVAIL (it was removed); marking it as OFFLINE provides *less* useful info (ie: the device can be offlined for various reasons). I think `zed` should not try to "outsmart" the user hiding potentially useful information.

Reply:
That's reasonable.  Reading back through the history of this change the original issue appears to have been devices might remain ONLINE even after being physically removed.  I could see this happening if the pool was idle.  The kernel modules wouldn't notice the missing device until it is next accessed.  The zed was updated to catch the udev remove event in user space and OFFLINE the drives.  It should have set the state to UNAVAIL or possibly even REMOVED which is a valid state.  There's no existing helper function to do this but it would be easy enough to add one.

Reply:
Yeah, from [man page](https://zfsonlinux.org/manpages/0.8.4/man8/zpool.8.html) the correct device status should be REMOVED:

> The device was physically removed while the system was running. Device removal detection is hardware-dependent and may not be supported on all platforms.

However, the `zpool_vdev_remove` function is about the newly introduced permanent device removal, and I feel as using the REMOVED state can lead to confusion. Moreover, without `zed` running, any IO to the removed device would cause ZFS to mark the device as UNAVAIL, so I think we should stick with that event for now.

To let `zed` immediately report the missing disk, should we issue a single call to `zpool_reopen_one`? It should immediately discover the missing device as UNAVAIL. We can even first OFFLINE the device and the reopen the pool: the device will be re-discovered as UNAVAIL.

Reply:
I just also tried to OFFLINE, then ONLINE a remove device. After the ONLINE, it become UNAVAIL. So, if we really want ZFS immediately detect a disconnected device, we can: a) reopen the pool or b) offline, then online the affected device. Which would be the preferred approach? That said, I am not sure we want to play reopen/offline/online anything just to let ZFS immediately see a disconnetect device at all, as any write will immediately notify ZFS of the missing disk.

Reply:
Actually, I was thinking something a little more like this 464909c6e0db19957fea54f817bc01680bc54b09 to transition the vdev in to the VDEV_STATE_REMOVE state (compiles,but otherwise untested).  This should provide [the same behavior](https://github.com/openzfs/zfs/blob/master/module/os/linux/zfs/vdev_disk.c#L755..L759) as intended by the kmods when they determine that the device has been removed.  There's a fair bit of logic in the kmods for specially this case which I think we should try and take advantage of.  If there are complications we can always adapt this to instead leave the device in the VDEV_STATE_CANT_OPEN state.  I think a reasonable case could be made for either.

Reply:
It somewhat worries me that the changes to mark the device as removed are not completely contained into `zed` - they extend to other part of ZFS itself. I would not make such changes for a 0.8.x point release. What do you think is the best method to set the device in VDEV_STATE_CANT_OPEN state? A pool reload? An offline/online step? Do we have a specific `zpool_vdev_` helper function? Thanks.

Reply:
I tend to agree, it's a larger change than I'd like to make to the 0.8 release as well.  Your current patch which restores the 0.7 behavior would be the safest course.  Both reopening the pool and onlining/offlining the failed vdev may have additional side effects like additional zpool events, restarting TRIM, etc so I'm not crazy about either approach.  We unfortunately don't have any existing `zpool_vdev_*` helper function to do exactly what we need.

I'd suggest we apply your fix as it stands in the PR to master (with the test case fixed as well).  This can then be ported to 0.8.  After this is merged I can open a new PR which makes the VDEV_STATE_REMOVED changed for the 2.0 release.